packages/r2/src/bucket.ts

```diff
@@ -0,0 +1,679 @@
+          400,
+          "include values must be httpMetadata and/or customMetadata strings."
+        );
+      }
+    }
+  }
+}
+
+export interface InternalR2BucketOptions {
+  blockGlobalAsyncIO?: boolean;
+}
+
+export class R2Bucket {
+  readonly #storage: Storage;
+  readonly #blockGlobalAsyncIO: boolean;
+
+  constructor(
+    storage: Storage,
+    { blockGlobalAsyncIO = false }: InternalR2BucketOptions = {}
+  ) {
+    this.#storage = storage;
+    this.#blockGlobalAsyncIO = blockGlobalAsyncIO;
+  }
+
+  #prepareCtx(method: Method, key?: string): RequestContext | undefined {
+    if (this.#blockGlobalAsyncIO) assertInRequest();
+    const ctx = getRequestContext();
+    ctx?.incrementInternalSubrequests();
+    // noinspection SuspiciousTypeOfGuard
+    if (method !== "LIST" && typeof key !== "string") {
+      throw new TypeError(
+        `Failed to execute '${method.toLowerCase()}'` +
+          " on 'R2Bucket': parameter 1 is not of type 'string'."
+      );
+    }
+
+    return ctx;
+  }
+
+  async head(key: string, ctx?: RequestContext): Promise<R2Object | null> {
+    if (ctx === undefined) ctx = this.#prepareCtx("HEAD", key);
+
+    // Validate key
+    validateKey("HEAD", key);
+
+    // Get value, returning null if not found
+    const stored = await this.#storage.headMaybeExpired?.<R2ObjectMetadata>(
+      key
+    );
+    // fix dates
+    await waitForOpenInputGate();
+    ctx?.advanceCurrentTime();
+    if (stored?.metadata === undefined) return null;
+    const { metadata } = stored;
+    parseR2ObjectMetadata(metadata);
+
+    return new R2Object(metadata);
+  }
+
+  /**
+   * Returns R2Object on a failure of the conditional specified in onlyIf.
+   */
+  async get(key: string): Promise<R2ObjectBody | null>;
+  async get(
+    key: string,
+    options: R2GetOptions
+  ): Promise<R2ObjectBody | R2Object | null>;
+  async get(
+    key: string,
+    options?: R2GetOptions
+  ): Promise<R2ObjectBody | R2Object | null> {
+    const ctx = this.#prepareCtx("GET", key);
+    options = options ?? {};
+    const { range = {} } = options;
+
+    // Validate key
+    validateKey("GET", key);
+    // Validate options
+    validateGetOptions(options);
+
+    // In the event that an onlyIf precondition fails, we return
+    // the R2Object without the body. Otherwise return with body.
+    const onlyIf = parseOnlyIf(options.onlyIf);
+    const meta = await this.head(key, ctx);
+    // if bad metadata, return null
+    if (meta === null) return null;
+    // test conditional should it exist
+    if (!testR2Conditional(onlyIf, meta) || meta?.size === 0) {
+      return new R2Object(meta);
+    }
+
+    let stored:
+      | R2StoredValueMeta<R2ObjectMetadata>
+      | RangeStoredValueMeta<R2ObjectMetadata>
+      | undefined;
+
+    // get data dependent upon whether suffix or range exists
+    if (typeof range.suffix === "number") {
+      try {
+        stored = await this.#storage.getSuffixMaybeExpired?.<R2ObjectMetadata>(
```

Comment:
Right below these attempts to get stored we have:
```js
if (stored?.metadata === undefined) return null;
```
Since only file-storage can actually handle these function calls, I assumed we should just return null as if the data does not exist. But maybe we can throw an error instead to let users know they have to use persistance for R2 to work? Open to suggestions.